[Gateway] Implement StakeGateway Message and Add Tests

h5law opened the pull request:
## Summary

### Human Summary

- Add stake field to the stake and gateway proto messages
- Update validation
- Update and add unit tests relating to
  - genesis
  - cli
  - staking
- Hydrate the stake keeper method with the correct logic
- Add makefile helpers

## Issue

Fixes (in part) [19](https://github.com/pokt-network/poktroll/issues/19)

## Type of change

Select one or more:

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

- [x] **Run all unit tests**: `make test_all_unit`
- [x] **Verify Localnet manually**: See the instructions [here](TODO: add link to instructions)

## Sanity Checklist

- [x] I have tested my changes using the available tooling
- [x] I have performed a self-review of my own code
- [x] I have commented my code, updated documentation and left TODOs throughout the codebase

Olshansk commented on x/gateway/client/cli/tx_stake_gateway.go:
```diff
@@ -3,29 +3,35 @@ package cli
 func CmdStakeGateway() *cobra.Command {
 	cmd := &cobra.Command{
-		Use:   "stake-gateway",
+		Use:   "stake-gateway [amount]",
```
See #59 for additions I made to usage & such

Olshansk commented on x/gateway/client/cli/tx_stake_gateway.go:
```diff
@@ -3,29 +3,35 @@ package cli
 				return err
 			}
-
+			stakeAmountString := args[0]
```
s/stakeAmountString/stakeString
s/stakeAmount/stake

It includes both amount & denom

Olshansk commented on x/gateway/client/cli/tx_stake_gateway_test.go:
```diff
@@ -0,0 +1,127 @@
+	}
+
+	// Initialize the Gateway Account by sending it some funds from the validator account that is part of genesis
+	sendArgs := []string{
```
See #59 for changes I made to leverage a helper.

Olshansk commented on x/gateway/keeper/msg_server_stake_gateway.go:
```diff
@@ -3,15 +3,81 @@ package keeper
+	if msg.Stake == nil {
+		return errors.Wrapf(types.ErrGatewayInvalidStake, "stake amount cannot be nil")
+	}
+	if gateway.Stake.IsGTE(*msg.Stake) {
```
@bryanchriswhite suggested use IsLTE here instead for readability - I agree.

Olshansk commented on x/gateway/types/errors.go:
```diff
@@ -1,12 +1,13 @@
 // x/gateway module sentinel errors
 var (
-	ErrSample = sdkerrors.Register(ModuleName, 1100, "sample error")
+	ErrSample                = errors.Register(ModuleName, 1100, "sample error")
```
Should we remove the sample?

Olshansk commented on x/gateway/types/genesis.go:
```diff
@@ -19,15 +21,32 @@ func DefaultGenesis() *GenesisState {
+		}
+		stakeAmount, err := sdk.ParseCoinNormalized(elem.Stake.String())
+		if !stakeAmount.IsValid() {
+			return fmt.Errorf("invalid stake amount for gateway %v; (%v)", elem.Stake, stakeAmount.Validate())
```
See #59 and return proper errors.

Olshansk reviewed (changes requested):
Overall LGTM, but I think the big piece is just catching up with the changes I made in the other PR.

Olshansk commented:
@h5law Also note that this is not "Hydration," but this is an implementation of the StakeGateway message. I recommend asking ChatGPT about the difference if you want to learn more.

![Screenshot 2023-10-17 at 5 03 20 PM](https://github.com/pokt-network/poktroll/assets/1892194/aecf1c61-0674-4801-9eee-8d338c02ce18)

h5law commented:
@Olshansk have addressed the comments and caught up with #59 only thing I removed one of your helpers in the network test utils, I couldn't get it to work with the types. It only worked for the application genesis state and I couldn't figure out a good way to make it generic without stepping into encoding and decoding `any` types so thought best this be put into the helpers that call this method so simplicities sake.

Olshansk commented on cmd/pocketd/cmd/root.go:
```diff
@@ -39,8 +39,6 @@ import (
 	"github.com/spf13/cobra"
 	"github.com/spf13/pflag"
 
-	// this line is used by starport scaffolding # root/moduleImport
```
Please add this back

Olshansk commented on x/application/client/cli/helpers_test.go:
```diff
@@ -24,6 +26,8 @@ func networkWithApplicationObjects(t *testing.T, n int) (*network.Network, []typ
 	cfg := network.DefaultConfig()
 	appGenesisState := network.DefaultApplicationModuleGenesisState(t, n)
-	network.HydrateGenesisState(t, &cfg, appGenesisState, types.ModuleName)
+	buf, err := cfg.Codec.MarshalJSON(appGenesisState)
```
If we're not going to use `HydrateGenesisState`, let's remove that funciton.

h5law replied:
Pretty sure I did

Olshansk reviewed (approved):
LGTM but PTAL at the two tiny comments before merging.
